adapter/src/hermes/adapter/stdio/stdio.cc

```diff
@@ -0,0 +1,993 @@
+  int ret = -1;
+  if (hermes::adapter::IsTracked(stream)) {
+    auto mdm = hermes::adapter::Singleton<MetadataManager>::GetInstance();
+    auto existing = mdm->Find(stream);
+    if (existing.second) {
+      LOG(INFO) << "Intercept fgetc." << std::endl;
+      unsigned char value;
+      auto ret_size = read_internal(existing, &value, sizeof(value), stream);
+      if (ret_size == 1) {
+        ret = value;
+      }
+    } else {
+      MAP_OR_FAIL(fgetc);
+      ret = real_fgetc_(stream);
+    }
+  } else {
+    MAP_OR_FAIL(fgetc);
+    ret = real_fgetc_(stream);
+  }
+  return (ret);
+}
+
+int HERMES_DECL(getc)(FILE *stream) {
+  int ret = -1;
+  if (hermes::adapter::IsTracked(stream)) {
+    auto mdm = hermes::adapter::Singleton<MetadataManager>::GetInstance();
+    auto existing = mdm->Find(stream);
+    if (existing.second) {
+      LOG(INFO) << "Intercept getc." << std::endl;
+      unsigned char value;
+      auto ret_size = read_internal(existing, &value, sizeof(value), stream);
+      if (ret_size == 1) {
+        ret = value;
+      }
+    } else {
+      MAP_OR_FAIL(fgetc);
+      ret = real_fgetc_(stream);
+    }
+  } else {
+    MAP_OR_FAIL(fgetc);
+    ret = real_fgetc_(stream);
+  }
+  return (ret);
+}
+
+/* NOTE: stdio.h typically implements getc() as a macro pointing to _IO_getc */
+int HERMES_DECL(_IO_getc)(FILE *stream) {
+  int ret;
+  if (hermes::adapter::IsTracked(stream)) {
+    auto mdm = hermes::adapter::Singleton<MetadataManager>::GetInstance();
+    auto existing = mdm->Find(stream);
+    if (existing.second) {
+      LOG(INFO) << "Intercept _IO_getc." << std::endl;
+      unsigned char value;
+      auto ret_size = read_internal(existing, &value, sizeof(value), stream);
+      if (ret_size == 1) {
+        ret = value;
+      }
+    } else {
+      MAP_OR_FAIL(_IO_getc);
+      ret = real__IO_getc_(stream);
+    }
+  } else {
+    MAP_OR_FAIL(_IO_getc);
+    ret = real__IO_getc_(stream);
+  }
+  return (ret);
+}
+
+/* NOTE: stdio.h typically implements putc() as a macro pointing to _IO_putc */
+int HERMES_DECL(_IO_putc)(int c, FILE *stream) {
+  int ret;
+  if (hermes::adapter::IsTracked(stream)) {
+    auto mdm = hermes::adapter::Singleton<MetadataManager>::GetInstance();
+    auto existing = mdm->Find(stream);
+    if (existing.second) {
+      LOG(INFO) << "Intercept _IO_putc char:" << (char *)&c << "." << std::endl;
+      ret = write_internal(existing, &c, 1, stream);
+    } else {
+      MAP_OR_FAIL(_IO_putc);
+      ret = real__IO_putc_(c, stream);
+    }
+  } else {
+    MAP_OR_FAIL(_IO_putc);
+    ret = real__IO_putc_(c, stream);
+  }
+  return (ret);
+}
+
+int HERMES_DECL(getw)(FILE *stream) {
+  int ret;
+  if (hermes::adapter::IsTracked(stream)) {
+    auto mdm = hermes::adapter::Singleton<MetadataManager>::GetInstance();
+    auto existing = mdm->Find(stream);
+    if (existing.second) {
+      LOG(INFO) << "Intercept getw." << std::endl;
+      unsigned char value;
```

Comment:
Isn't `getw` supposed to read an `int`?

Reply:
With int the char conversion fails with undefined behavior. Unsigned char gives most consistent results.

Reply:
But you're only reading 1 byte. It's supposed to read `sizeof(int)` bytes.

Reply:
Ok, i tried that. It worked.